### PR TITLE
Send service background/terminate lifecycle events from mobile apps

### PR DESCRIPTION
**Pull request by buoyad**
Gives the service an opportunity to request more execution time on iOS if it is e.g. sending a message and adds a new `AppState` for when the app is about to be terminated. r? @keybase/react-hackers for java/objC and @mmaxim for go

### CODE REVIEW

**Comment by mmaxim** [comment]
Are you sure this is going to play out exactly like it did with the JS?

File: shared/react-native/ios/Keybase/AppDelegate.m [hunk 1]
```diff
@@ -229,15 +232,49 @@ - (void)applicationDidEnterBackground:(UIApplication *)application {
...
+      KeybaseSetAppStateForeground();
+      break;
+    case UIApplicationStateBackground:
+      KeybaseSetAppStateBackground();
+      break;
+    case UIApplicationStateInactive:
+      KeybaseSetAppStateInactive();
+      break;
+    default:
+      KeybaseSetAppStateForeground();
```
Comment (buoyad) [inline shared/react-native/ios/Keybase/AppDelegate.m]: this is to match the behavior on [this line](https://github.com/keybase/client/pull/11142/files#diff-25ab83d9513d1ae61c1f3a4d40b0f15fL22)

**Comment by buoyad** [comment]
It should, I compared it against the native modules for [iOS](https://github.com/facebook/react-native/blob/master/React/Modules/RCTAppState.m) and [android](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/appstate/AppStateModule.java) to send the same app states.

File: shared/actions/app.js [hunk 2]
```diff
@@ -1,40 +1,14 @@
...
-  const state =
-    {
-      active: RPCTypes.appStateAppState.foreground,
-      background: RPCTypes.appStateAppState.background,
-      inactive: RPCTypes.appStateAppState.inactive,
-    }[nextAppState] || RPCTypes.appStateAppState.foreground
-  logger.info(`setting app state on service to: ${state}`)
-
-  return Saga.sequentially([
-    Saga.put(AppGen.createChangedFocus({appFocused})),
```
Comment (chrisnojima) [inline shared/actions/app.js]: we need to make sure removing this doesn't break anything right

**Comment by chrisnojima** [comment]
Additionally this should likely be removed: https://github.com/keybase/client/blob/dd44fd182907247873548b0968eadf33b7399cb1/shared/actions/push.js#L135 but we need to validate the native calls go through correctly first

File: shared/actions/app.js (hunk 2 above)
Comment (buoyad) [reply shared/actions/app.js]: oof yeah, it's really just the RPC that should be removed here. thanks

**Comment by buoyad** [comment]
iOS log produced by having both native & JS setAppState enabled:

```
2018-04-02T17:36:24.041207-04:00 ▶ [DEBU keybase keybase.go:285] 9af DANNYDEBUG GOT: INACTIVE
2018-04-02T17:36:24.063772-04:00 ▶ [DEBU keybase appstate.go:29] 9be DANNYDEBUG GOT (FROM JS): INACTIVE
2018-04-02T17:36:24.590901-04:00 ▶ [DEBU keybase keybase.go:303] 9c2 DANNYDEBUG GOT: APPDIDENTERBACKGROUND
2018-04-02T17:36:24.590953-04:00 ▶ [DEBU keybase keybase.go:281] 9c3 DANNYDEBUG GOT: BACKGROUND
2018-04-02T17:36:24.608058-04:00 ▶ [DEBU keybase appstate.go:29] 9ce DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T17:36:29.841349-04:00 ▶ [DEBU keybase keybase.go:281] 9eb DANNYDEBUG GOT: BACKGROUND
2018-04-02T17:36:30.130880-04:00 ▶ [DEBU keybase keybase.go:277] 9fb DANNYDEBUG GOT: FOREGROUND
2018-04-02T17:36:30.141476-04:00 ▶ [DEBU keybase appstate.go:29] a1b DANNYDEBUG GOT (FROM JS): FOREGROUND
2018-04-02T17:36:40.691711-04:00 ▶ [DEBU keybase keybase.go:285] ad2 DANNYDEBUG GOT: INACTIVE
2018-04-02T17:36:40.700986-04:00 ▶ [DEBU keybase appstate.go:29] ae3 DANNYDEBUG GOT (FROM JS): INACTIVE
2018-04-02T17:36:41.242579-04:00 ▶ [DEBU keybase keybase.go:303] ae7 DANNYDEBUG GOT: APPDIDENTERBACKGROUND
2018-04-02T17:36:41.242630-04:00 ▶ [DEBU keybase keybase.go:281] ae8 DANNYDEBUG GOT: BACKGROUND
2018-04-02T17:36:41.258812-04:00 ▶ [DEBU keybase appstate.go:29] af3 DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T17:36:50.858069-04:00 ▶ [DEBU keybase keybase.go:281] afa DANNYDEBUG GOT: BACKGROUND
2018-04-02T17:36:51.146755-04:00 ▶ [DEBU keybase keybase.go:277] b01 DANNYDEBUG GOT: FOREGROUND
2018-04-02T17:36:51.157528-04:00 ▶ [DEBU keybase appstate.go:29] b21 DANNYDEBUG GOT (FROM JS): FOREGROUND
2018-04-02T17:36:53.268877-04:00 ▶ [DEBU keybase keybase.go:285] bc6 DANNYDEBUG GOT: INACTIVE
2018-04-02T17:36:53.279318-04:00 ▶ [DEBU keybase appstate.go:29] bd5 DANNYDEBUG GOT (FROM JS): INACTIVE
2018-04-02T17:36:53.827095-04:00 ▶ [DEBU keybase keybase.go:303] bd9 DANNYDEBUG GOT: APPDIDENTERBACKGROUND
2018-04-02T17:36:53.827149-04:00 ▶ [DEBU keybase keybase.go:281] bda DANNYDEBUG GOT: BACKGROUND
2018-04-02T17:36:53.839028-04:00 ▶ [DEBU keybase appstate.go:29] be5 DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T17:36:58.079361-04:00 ▶ [DEBU keybase keybase.go:277] 283 DANNYDEBUG GOT: FOREGROUND
2018-04-02T17:37:01.740362-04:00 ▶ [DEBU keybase keybase.go:285] 662 DANNYDEBUG GOT: INACTIVE
2018-04-02T17:37:01.753397-04:00 ▶ [DEBU keybase appstate.go:29] 671 DANNYDEBUG GOT (FROM JS): INACTIVE
2018-04-02T17:37:02.291833-04:00 ▶ [DEBU keybase keybase.go:303] 675 DANNYDEBUG GOT: APPDIDENTERBACKGROUND
2018-04-02T17:37:02.291887-04:00 ▶ [DEBU keybase keybase.go:281] 676 DANNYDEBUG GOT: BACKGROUND
2018-04-02T17:37:02.309440-04:00 ▶ [DEBU keybase appstate.go:29] 683 DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T17:37:05.371370-04:00 ▶ [DEBU keybase keybase.go:281] 68a DANNYDEBUG GOT: BACKGROUND
2018-04-02T17:37:05.658613-04:00 ▶ [DEBU keybase keybase.go:277] 691 DANNYDEBUG GOT: FOREGROUND
2018-04-02T17:37:05.670408-04:00 ▶ [DEBU keybase appstate.go:29] 6b1 DANNYDEBUG GOT (FROM JS): FOREGROUND
```
the `AppDidEnterBackground` calls are the ones that can return `true` to request more time. The extra `Background` notifications are from the iOS `appWillEnterForeground` hook, which RN claims to handle but it looks like it doesn't (this is sent when the state is already 'background' so should be a no-op). Otherwise, it looks like the behavior is the same, with native notifications coming a bit earlier. Testing android now.

**Comment by buoyad** [comment]
Android is even more 1-to-1
```
2018-04-02T21:57:20.697524Z ▶ [DEBU keybase appstate.go:29] ce1 DANNYDEBUG GOT (FROM JS): FOREGROUND
2018-04-02T21:57:21.286059Z ▶ [DEBU keybase keybase.go:281] cf8 DANNYDEBUG GOT: BACKGROUND
2018-04-02T21:57:21.395599Z ▶ [DEBU keybase appstate.go:29] d0c DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T21:57:22.421292Z ▶ [DEBU keybase keybase.go:296] d13 DANNYDEBUG GOT: APPWILLEXIT
2018-04-02T21:57:24.356732Z ▶ [DEBU keybase keybase.go:277] 0dc DANNYDEBUG GOT: FOREGROUND
2018-04-02T21:57:26.169964Z ▶ [DEBU keybase appstate.go:29] 2ee DANNYDEBUG GOT (FROM JS): FOREGROUND
2018-04-02T21:57:27.115020Z ▶ [DEBU keybase keybase.go:281] 3a5 DANNYDEBUG GOT: BACKGROUND
2018-04-02T21:57:27.264432Z ▶ [DEBU keybase appstate.go:29] 3b9 DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T21:57:28.370430Z ▶ [DEBU keybase keybase.go:277] 3c3 DANNYDEBUG GOT: FOREGROUND
2018-04-02T21:57:28.494992Z ▶ [DEBU keybase appstate.go:29] 402 DANNYDEBUG GOT (FROM JS): FOREGROUND
2018-04-02T21:57:30.319098Z ▶ [DEBU keybase keybase.go:281] 482 DANNYDEBUG GOT: BACKGROUND
2018-04-02T21:57:30.444492Z ▶ [DEBU keybase appstate.go:29] 496 DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T21:57:31.352254Z ▶ [DEBU keybase keybase.go:277] 49d DANNYDEBUG GOT: FOREGROUND
2018-04-02T21:57:31.544343Z ▶ [DEBU keybase appstate.go:29] 545 DANNYDEBUG GOT (FROM JS): FOREGROUND
2018-04-02T21:57:32.751362Z ▶ [DEBU keybase keybase.go:281] 55c DANNYDEBUG GOT: BACKGROUND
2018-04-02T21:57:32.848107Z ▶ [DEBU keybase appstate.go:29] 570 DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T21:57:33.855468Z ▶ [DEBU keybase keybase.go:277] 577 DANNYDEBUG GOT: FOREGROUND
2018-04-02T21:57:33.980754Z ▶ [DEBU keybase appstate.go:29] 61f DANNYDEBUG GOT (FROM JS): FOREGROUND
2018-04-02T21:57:34.754196Z ▶ [DEBU keybase keybase.go:281] 638 DANNYDEBUG GOT: BACKGROUND
2018-04-02T21:57:34.836930Z ▶ [DEBU keybase appstate.go:29] 64e DANNYDEBUG GOT (FROM JS): BACKGROUND
2018-04-02T21:57:35.658174Z ▶ [DEBU keybase keybase.go:296] 655 DANNYDEBUG GOT: APPWILLEXIT
2018-04-02T21:57:37.761532Z ▶ [DEBU keybase keybase.go:277] 1db DANNYDEBUG GOT: FOREGROUND
2018-04-02T21:57:39.153224Z ▶ [DEBU keybase appstate.go:29] 2ee DANNYDEBUG GOT (FROM JS): FOREGROUND
```

The only extra call the service sometimes gets is `AppWillExit`.

**Comment by buoyad** [comment]
It looks like the only difference is on iOS: an extra `background` call before foregrounding the app (from iOS's `appWillEnterForeground`). @chrisnojima @mmaxim I think they'll do the same thing, how does this look?

**Pull request merged**